src/compound.rs

```diff
@@ -4,183 +4,115 @@
 //
 // Copyright (c) DUSK NETWORK. All rights reserved.
 
-use core::marker::PhantomData;
+use rkyv::{Archive, Deserialize};
 
-use rkyv::{Archive, Deserialize, Infallible};
-
-use crate::annotations::{ARef, Annotation};
 use crate::link::{ArchivedLink, Link};
-use crate::{AWrap, Branch, BranchMut, Walker};
+use crate::{Annotation, Branch, BranchMut, MaybeArchived, Store, Walker};
 
 /// The response of the `child` method on a `Compound` node.
-pub enum Child<'a, C, A>
+pub enum Child<'a, C, A, S>
 where
-    C: Compound<A>,
-    A: Annotation<C::Leaf>,
+    S: Store,
+    C: Compound<A, S>,
+    C::Leaf: Archive,
 {
     /// Child is a leaf
     Leaf(&'a C::Leaf),
     /// Child is an annotated subtree node
-    Node(&'a Link<C, A>),
+    Link(&'a Link<C, A, S>),
     /// Empty slot
     Empty,
     /// No more children
-    EndOfNode,
+    End,
 }
 
 /// The response of the `child` method on a `Compound` node.
-pub enum ArchivedChild<'a, C, A>
+pub enum ArchivedChild<'a, C, A, S>
 where
-    C: Compound<A>,
-    C::Leaf: Archive,
-    A: Annotation<C::Leaf>,
+    S: Store,
+    C: Compound<A, S>,
 {
     /// Child is a leaf
     Leaf(&'a <C::Leaf as Archive>::Archived),
     /// Child is an annotated subtree node
-    Node(&'a ArchivedLink<C, A>),
+    Link(&'a ArchivedLink<C, A, S>),
     /// Empty slot
     Empty,
     /// No more children
-    EndOfNode,
+    End,
 }
 
 /// The response of the `child_mut` method on a `Compound` node.
-#[derive(Debug)]
-pub enum ChildMut<'a, C, A>
+pub enum ChildMut<'a, C, A, S>
 where
-    C: Compound<A>,
-    A: Annotation<C::Leaf>,
+    S: Store,
+    C: Compound<A, S>,
 {
     /// Child is a leaf
     Leaf(&'a mut C::Leaf),
     /// Child is an annotated node
-    Node(&'a mut Link<C, A>),
+    Link(&'a mut Link<C, A, S>),
     /// Empty slot
     Empty,
     /// No more children
-    EndOfNode,
+    End,
 }
 
 /// Trait to support branch traversal in archived nodes
-pub trait ArchivedCompound<C, A>
+pub trait ArchivedCompound<C, A, S>
 where
-    C: Compound<A>,
+    S: Store,
+    C: Compound<A, S>,
     C::Leaf: Archive,
-    A: Annotation<C::Leaf>,
 {
     /// Returns an archived child
-    fn child(&self, ofs: usize) -> ArchivedChild<C, A>;
-
-    /// Constructs a branch from this root compound
-    fn walk<'a, W>(&'a self, walker: W) -> Option<Branch<'a, C, A>>
-    where
-        W: Walker<C, A>,
-        C: Archive<Archived = Self>,
-    {
-        Branch::walk(AWrap::Archived(self), walker)
-    }
+    fn child(&self, ofs: usize) -> ArchivedChild<C, A, S>;
 }
 
 /// A type that can recursively contain itself and leaves.
-pub trait Compound<A>: Sized
-where
-    A: Annotation<Self::Leaf>,
-{
+pub trait Compound<A, S>: Sized + Archive {
     /// The leaf type of the Compound collection
-    type Leaf;
+    type Leaf: Archive;
 
     /// Get a reference to a child    
-    fn child(&self, ofs: usize) -> Child<Self, A>;
+    fn child(&self, ofs: usize) -> Child<Self, A, S>
```

Comment:
It was not added in this PR, so what follows is only a suggestion/comment - I think it'd be simpler if child method returned a Compound type, and what is now Child enum would be treated as "this node" enum. This way we could eliminate the Leaf type variable. It is confusing that not every node is a Compound, and that there is a difference between "this" which is a Compound, and a child, which is a Child. I'd make everything a Compound and rename Child to Node and keep it as sth like "current" or "this" inside a compound (or merge Child (Node) with Compound).

Reply:
i'm not sure if i understand what you mean here.